tidy3d/components/data/monitor_data.py

```diff
@@ -133,7 +133,136 @@ def symmetry_expanded_copy(self) -> AbstractFieldData:
         return self.copy(update=update_dict)
 
 
-class FieldData(FieldDataset, AbstractFieldData):
+class ElectromagneticFieldData(AbstractFieldData, ElectromagneticFieldDataset, ABC):
+    """Collection of electromagnetic fields."""
+
+    @property
+    def _tangential_dims(self) -> List[str]:
+        """For a 2D monitor data, return the names of the tangential dimensions. Raise if cannot
+        confirm that the associated monitor is 2D."""
+        zero_dims = np.where(np.array(self.monitor.size) == 0)[0]
+        if zero_dims.size != 1:
+            raise DataError("Data must be 2D to get tangential dimensions.")
+        tangential_dims = ["x", "y", "z"]
+        tangential_dims.pop(zero_dims[0])
+
+        return tangential_dims
+
+    @property
+    def _tangential_fields(self) -> Dict[str, DataArray]:
+        """For a 2D monitor data, return a dictionary with only the tangential fields. Raise if any
+        of the tangential field components is missing."""
+        tan_dims = self._tangential_dims
+        field_components = ["E" + dim for dim in tan_dims] + ["H" + dim for dim in tan_dims]
+        for field in field_components:
+            if field not in self.field_components:
+                raise DataError(f"Tangential field component {field} is missing in data.")
+        normal_dim = "xyz"[self.monitor.size.index(0)]
+        return {field: getattr(self, field).squeeze(dim=normal_dim) for field in field_components}
+
+    @cached_property
+    def _plane_grid_boundaries(self) -> Tuple[ArrayLike[float, 1], ArrayLike[float, 1]]:
+        """For a 2D monitor data, return the boundaries of the in-plane grid from the stored field
+        coordinates."""
+        tan_dims = self._tangential_dims
+        tan_fields = self._tangential_fields
+        plane_bounds0 = tan_fields["E" + tan_dims[1]].coords[tan_dims[0]].values
+        plane_bounds1 = tan_fields["E" + tan_dims[0]].coords[tan_dims[1]].values
+        return plane_bounds0, plane_bounds1
+
+    @cached_property
+    def _diff_area(self) -> xr.DataArray:
```

Comment:
So this technically includes the data just outside of the monitor volume right? The points we use to interpolate into the volume? Therefore, the sizes are actually not equal to the intersection volume between the simulation and the FeldMonitor, like the user might expect.  Is it ok?

Reply:
Yeah very good point. I modified the sizes such that the integration now happens exactly on the analytic monitor plane. I also added more backend tests to check that an analytically air-tight box which is not commensurate with the cell produces flux very close to unity when computed in this way.

The only caveat is that server-side flux computation still snaps monitors to centers and to a discrete number of pixels. I gave this another thought but it is unfortunately not so straightforward to fix this (low reward/effort ratio...). So I added a note about this, that the flux computed in these two different ways will only match exactly if the FieldMonitor is placed at a cell center and spans an integer number of cells.

Reply:
hmm ok so the client-side flux will actually be more accurate than the server-side flux in the general case (non-integer number of cells in the monitor volume)? Seems potentially problematic or limiting, but I guess this is a separate PR. I think most users will assume the flux is computed using the monitor volume and could be confused.